Added shared memory support usage description

MatthiasKillat opened the pull request:
Signed-off-by: Matthias Killat <matthias.killat@apex.ai>

Add a brief description how to enable shared memory support. The purpose is to enable a ROS2 user to activate the Shared Memory feature in her applications when using rmw_cyclonedds. 

Unfortunately there is no (simple) example yet in the ROS default examples where Shared Memory is actually used (at least none that I am aware of). However, it is fairly straightforward how the talker and listener example needs to be adapted to demonstrate use of shared memory support (essentially just use a fixed size data type, i.e. not a string). This is briefly sketched as well as a list of current restrictions.

Most of the important restrictions should be covered but some of them might change slightly (e.g. QoS requirements) and the document will need updates in this case.

Closes #304

MatthiasKillat commented on shared_memory_support.md:
````diff
@@ -0,0 +1,218 @@
+ros2 run demo_nodes_cpp listener
+```
+
+If the Shared Memory configration was successfully activated and recognized, you should get some output containing
````
configuration

FerdinandSpitzschnueffler commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+## Zero-Copy Publish-Subscribe Communication
+
+Whether true zero-copy publish subscribe communication via Shared Memory is possible depends on several factors. Here zero copy means no copy or serialization is performed.
```
zero-copy

FerdinandSpitzschnueffler commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+If the Shared Memory transfer conditions are met, publish will internally loan a Shared Memory chunk from iceoryx and copy the message payload into it without the serialization required for network transfer. Any connected subscription has read-only access to this message data.
+
+While this API will not allow true zero-copy transfer, it still will improve performance for sufficiently large message sizes since it bypasses the loopback interface and does not perform serialization. The actual size where it will outoperform the loopback interface primarily depends on the actual hardware and system load.
```
typo: outperform

FerdinandSpitzschnueffler commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+### Local Subscription
+
+Only local subscriptions on the same machine will receive data via Shared Memory and RouDi is required to enable this. Remote subscriptions will always receive the data via the network interface. There can only be one Roudi process be running per machine.
```
Please remove one `be` in the last sentence.

FerdinandSpitzschnueffler commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+### Iceoryx Shared Memory Configuration
+
+Iceoryx uses configurable memory pools to define different sizes of memory chunks that will be used to store messages in Shared Memory. These will be obtained when data is send via iceoryx, either when explicitly requested with the [Loan API](#Loan-API) or implicitly by the [Regular Publish API](#Regular-Publish-API).
```
is sent

FerdinandSpitzschnueffler commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+Iceoryx uses configurable memory pools to define different sizes of memory chunks that will be used to store messages in Shared Memory. These will be obtained when data is send via iceoryx, either when explicitly requested with the [Loan API](#Loan-API) or implicitly by the [Regular Publish API](#Regular-Publish-API).
+
+Depending on the size and frequency of messages send, the default configuration may not be sufficient to guarantee that memory can be loaned and hence the data send. In this case it might help to use a custom configuration for the shared memory pools to increase the available Shared Memory. The configuration options are described in the [iceoryx configuration guide](https://github.com/eclipse-ice ...
```
sent

clalancette commented on shared_memory_support.md:
```diff
@@ -0,0 +1,218 @@
+
+## Requirements
+
+Currently Shared Memory transport is only supported on Linux. It is available in the rmw_cyclonedds implementation used by the ROS2 Rolling or Galactic Geochelone distribution.
```
As a general statement in this document, we prefer "ROS 2" over "ROS2".

MatthiasKillat replied:
Fixed throughout the docuement.

MatthiasKillat commented:
I corrected some minor inaccuracies and also added a description how to use the iceoryx introspection to verify whether shared memory is actually used. The introspection works for this purpose but is not build by default, it will likely be updated in the future. It is still useful to get a deeper look into system performance and therefore I added it here.

I think this will be very useful to have in the Galactic release and even if there is always room for improvement we should merge the document to master and to galactic soon after.

I colleagues trying it and they found it useful to get started, so it should be in a decent state.

@eboasson @clalancette Let me know what you think how we can proceed here.

MatthiasKillat commented:
@eboasson @clalancette @ivanpauno Just a quick reminder that we would like to get this into the patch 1 of ROS 2 Galactic as far as I know so we should get this merged to master. Feel free to give feedback and I will fix the issues. 

We are already in the process of internally verifying that the patch will not break anything. @eboasson checked this on ARM64 and found no issues. @dkroenke, others and myself will do so for other platforms. Since this is just a text document it cannot invalidate the tests @eboasson already did.

MatthiasKillat commented:
@dkroenke @ivanpauno I corrected the mentioned issues and we should be good to go.

ivanpauno commented on shared_memory_support.md:
```diff
@@ -0,0 +1,287 @@
+Enabling Shared Memory imposes additional restrictions on publishers and subscriptions in ROS applications if Shared Memory is used for communication. In this case the data is received using an internal queue (per subscription).
+Note that depending on the QoS settings Shared Memory might not be used (cf. [Restrictions](#Restrictions)).
+
+1. *SubQueueCapacity* defines the maximum number of samples a subscription can locally store in Shared Memory. If more samples arrive via Shared Memory, the least recent ones will be discarded.
+2. *SubHistoryRequest* is the maximum number of samples a late joining subscription will request from a corresponding publisher.
+3. *PubHistoryCapacity* is the maximum number of samples a publisher will keep for late joining subscriptions.
+4. *Loglevel* controls the output of the iceoryx runtime and can be set to verbose, debug, info, warn, error, fatal and off in order of decreasing output level.
```
sounds good to me!

budrus commented:
@MatthiasKillat @ivanpauno @clalancette @eboasson As this is the shared memory user guide, it would be cool to get this merged so I do not have to send links to a PR. What is missing to have progress here?

clalancette commented on shared_memory_support.md:
```diff
@@ -0,0 +1,280 @@
+ROS 2 needs to be installed as described in [Installing ROS 2 Rolling](https://docs.ros.org/en/rolling/Installation/Ubuntu-Install-Binary.html).
+It can also be build from sources directly [Building ROS 2 Rolling](https://docs.ros.org/en/rolling/Installation/Ubuntu-Development-Setup.html). Using the ROS 2 Galactic installation as described in [Installing ROS 2 Galactic](https://docs.ros.org/en/galactic/Installation/Ubuntu-Install-Binary.html) and  [Building ROS 2 Galactic](https://docs.ros.org/en/galactic/Installation/Ubuntu-Development-Se ...
+
+In both cases rmw_cyclonedds is build with Shared Memory support by default.
```
```suggestion
In both cases rmw_cyclonedds is built with Shared Memory support by default.
```

clalancette reviewed (approved):
One nit, but looks OK to me.  I'm going to go ahead and fix that nit, then merge this.

eboasson commented:
@clalancette this seems to have dropped off the radar. I'll merge it now, but I don't know if that alone is enough to include it in the patch release.
